### ISSUE TITLE
commands.get_app_dir() is broken for development installs

### ISSUE THREAD

**Issue by mgeier** [jupyterlab/jupyterlab#3807]
If I install JupyterLab from a Git checkout with

    python3 -m pip install -e .

... and later run

    jupyter lab build

... I get an error (`Permission denied: '/usr/share/jupyter'`) because the logic in `commands.get_app_dir()` seems to be broken:

```python
if HERE.startswith(userbase) and not app_dir.startswith(userbase):
    app_dir = pjoin(userbase, 'share', 'jupyter', 'lab')
```

I guess this works for a non-development install because `HERE` will start with something like `$HOME/.local`, but it doesn't work for development installs because `HERE` can be anything, e.g. `/wherever/i/checked/out/jupyterlab/jupyterlab`.

**Comment by blink1073** [comment]
Hi @mgeier, what is `python -c "import site; print(site.getuserbase()); import jupyterlab; print(jupyterlab.__path__)`?

**Comment by mgeier** [comment]
With the closing double quotes its

```
/home/mg/.local
['/home/mg/git/jupyterlab/jupyterlab']
```

The last one is just where I happened to check out JupyterLab, this could of course be anything.

**Comment by blink1073** [comment]
I'm still not seeing the bug.  What does `python -c "from jupyterlab.commands import get_app_dir; print(get_app_dir())"` show?

**Comment by mgeier** [comment]
It shows

    /usr/share/jupyter/lab

... which doesn't exist on my system and `/usr/share` is not writable for my user (which makes sense).

It *should* show

    /home/mg/.local/share/jupyter/lab

... which is actually the Jupyter app dir.

> I'm still not seeing the bug.

The bug is that `HERE.startswith(userbase)` is required to be `True`, which it isn't for development installs.

**Comment by blink1073** [comment]
Ahh, I think the root of the issue that that your `sys.prefix` is at the root level.  We have detection for `pip install --user` installs, which is what the line you are referring to does, but not for development installs explicitly.  Just to verify, what is `python -c "import sys; print(sys.prefix)"`?

**Comment by blink1073** [comment]
We could explicitly check for a root level `sys.prefix` and only use it as the default app directory if we are also explicitly installed in one of the locations given by `site.getsitepackages()`, otherwise defaulting to `userbase`.

**Comment by mgeier** [comment]
My `sys.prefix` is

    /usr

I should probably add some more information. I'm using Debian Linux (testing a.k.a. Buster) and its standard Python installation.
The python binary is in `/usr/bin/python3.6` and `site.getsitepackages()` is

    ['/usr/local/lib/python3.6/dist-packages', '/usr/lib/python3/dist-packages', '/usr/lib/python3.6/dist-packages']

I'm also using the `pip` package from Debian, which (since not all too long) seems to install stuff into `$HOME/.local`, even if the `--user` option is not given.

So should my "app dir" now be `$HOME/.local/share/jupyter/lab` or some directory inside my Git checkout?

**Comment by blink1073** [comment]
Okay, everything is consistent based on your outputs.  The editable installed location is not one we were explicitly handling.  In the case of virtual or conda environments, the default behavior of using `sys.prefix` was fine.  We need additional logic to handle the case of an editable install in a system level `sys.prefix`.

**Comment by blink1073** [comment]
For now, you can set the `JUPYTERLAB_DIR` environment variable to `~/.local/share/jupyter/lab`.

**Comment by mgeier** [comment]
Thanks @blink1073, I can confirm that the work-around works for me:

    JUPYTERLAB_DIR=~/.local/share/jupyter/lab python3 -m jupyterlab build

It should be noted that the same environment variable also has to be defined when starting JupyterLab:

    JUPYTERLAB_DIR=~/.local/share/jupyter/lab python3 -m jupyterlab

**Comment by blink1073** [comment]
Right, I'd suggest adding it to `~/.bashrc` for now until the next release as `export JUPYTERLAB_DIR=~/.local/share/jupyter/lab`.

**Comment by mgeier** [comment]
Thanks, that works nicely!
